### PR TITLE
ihaskell: add module

### PR DESCRIPTION
**Pull request by srid**
- https://github.com/gibiansky/IHaskell
- https://begriffs.com/posts/2016-01-20-ihaskell-notebook.html

### CODE REVIEW

**Comment by srid** [comment]
cc @vaibhavsagar

File: modules/services/ihaskell.nix [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+  cfg = config.services.ihaskell;
+  ihaskellSrc = pkgs.fetchFromGitHub {
+    owner = "gibiansky";
+    repo = "IHaskell";
+    rev = "c2cb8e6789c3f2485ba9e1e0436177b700edb227";
+    sha256 = "0h5hpdfa3daabf1ijl7y5qnpy60yrf7h0ws2n3v5przfmgr8phv5";
+  };
+  ihaskellResult = (import (ihaskellSrc + /release.nix) {
+    compiler = "ghc864";
+    packages = self: cfg.extraPackages self;
```
Comment (vaibhavsagar) [inline modules/services/ihaskell.nix]: There are also `pythonPackages` and (badly named) `systemPackages` attributes. What do you think about making those configurable?

Comment (vaibhavsagar) [reply modules/services/ihaskell.nix]: Also, you can replace this with just `cfg.extraPackages` and it should work the same :smile:.

File: modules/services/ihaskell.nix [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+  cfg = config.services.ihaskell;
+  ihaskellSrc = pkgs.fetchFromGitHub {
+    owner = "gibiansky";
+    repo = "IHaskell";
+    rev = "c2cb8e6789c3f2485ba9e1e0436177b700edb227";
+    sha256 = "0h5hpdfa3daabf1ijl7y5qnpy60yrf7h0ws2n3v5przfmgr8phv5";
+  };
+  ihaskellResult = (import (ihaskellSrc + /release.nix) {
+    compiler = "ghc864";
```
Comment (vaibhavsagar) [inline modules/services/ihaskell.nix]: I don't think hardcoding the compiler version here like this is a great idea. I would prefer to have it be configurable, but default to `ghc84` (or the latest available version in nixpkgs).

File: modules/services/ihaskell.nix (hunk 1 above)
Comment (srid) [reply modules/services/ihaskell.nix]: Any reason someone would want to customize `pythonPackages` or `systemPackages` when using ihaskell?

Comment (vaibhavsagar) [reply modules/services/ihaskell.nix]: One example is using Z3 from IHaskell.

**Comment by srid** [comment]
@rycee it looks like `master` had some recent commits force-removed; should I update this branch to reflect that?

**Comment by rycee** [comment]
Hmm, force-removed? I haven't knowingly removed any commits, do you have any IDs of the removed commits?

**Comment by srid** [comment]
Yea see the commit history of `master` by the time I forked the repo: https://github.com/srid/home-manager/commits/master

**Comment by srid** [comment]
Oops; nevermind I got confused by the older date in the latest commit!

File: modules/services/ihaskell.nix [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+      description = "Autostart an IHaskell notebook service.";
+    };
+
+    compiler = mkOption {
+      default = "ghc864";
+      example = literalExample ''
+        ghc864
+      '';
+      description = ''
+        The Haskell compiler to use
```
Comment (rycee) [inline modules/services/ihaskell.nix]: Descriptions should end in a period.

File: modules/services/ihaskell.nix [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+{
+  meta.maintainers = [ maintainers.rycee ];
+
+  options.services.ihaskell = {
+    enable = mkOption {
+      default = false;
+      description = "Autostart an IHaskell notebook service.";
+    };
+
+    compiler = mkOption {
```
Comment (rycee) [inline modules/services/ihaskell.nix]: Missing type.

File: modules/services/ihaskell.nix [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+      default = "ghc864";
+      example = literalExample ''
+        ghc864
+      '';
+      description = ''
+        The Haskell compiler to use
+      '';
+    };
+
+    notebooksPath = mkOption {
```
Comment (rycee) [inline modules/services/ihaskell.nix]: Missing type.

File: modules/services/ihaskell.nix [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+      default = "$HOME/ihaskell";
+      example = literalExample ''
+        $HOME/projects/ihaskell-notebooks
+      '';
+      description = ''
+        Directory where iHaskell will store notebooks
```
Comment (rycee) [inline modules/services/ihaskell.nix]: The project name should be used consistently, in [their readme](https://github.com/gibiansky/IHaskell/blob/master/README.md) they seem to prefer IHaskell. So I would suggest to change iHaskell and ihaskell to IHaskell throughout the module.

File: modules/services/ihaskell.nix [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        ]
+      '';
+      description = ''
+        Extra packages available to ghc when running ihaskell. The
+        value must be a function which receives the attrset defined
+        in <varname>haskellPackages</varname> as the sole argument.
+      '';
+    };
+
+    pythonPackages = mkOption {
```
Comment (rycee) [inline modules/services/ihaskell.nix]: Seems unused.

File: modules/services/ihaskell.nix [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+          pythonPackages.numpy
+          pythonPackages.z3
+        ]
+      '';
+      description = ''
+        Extra Python packages available when running ihaskell.
+      '';
+    };
+    
+    systemPackages = mkOption {
```
Comment (rycee) [inline modules/services/ihaskell.nix]: Seems unused.

File: modules/services/ihaskell.nix [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+        Extra system packages available when running ihaskell.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.user.services.ihaskell = {
+      Unit = {
+        Description = "iHaskell notebook instance";
+        After = [ "network.target" ];
```
Comment (rycee) [inline modules/services/ihaskell.nix]: There is no `network.target` in the systemd user session.
```suggestion
        After = [ "graphical-session-pre.target" ];
```

File: modules/services/ihaskell.nix [hunk 10]
```diff
@@ -0,0 +1,113 @@
+# Adapted from Nixpkgs.
+
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  cfg = config.services.ihaskell;
+  ihaskellSrc = pkgs.fetchFromGitHub {
```
Comment (rycee) [inline modules/services/ihaskell.nix]: Why is this here? Is there no IHaskell package in Nixpkgs? If not then please add it to Nixpkgs and then refer to that package here. I'm very reluctant to add packages to Home Manager, they belong in Nixpkgs.

File: modules/modules.nix [hunk 11]
```diff
@@ -96,6 +96,7 @@ let
     (loadModule ./services/flameshot.nix { })
     (loadModule ./services/gnome-keyring.nix { })
     (loadModule ./services/gpg-agent.nix { })
+    (loadModule ./services/ihaskell.nix { })
```
Comment (rycee) [inline modules/modules.nix]: Since this module uses systemd it should be limited to Linux:
```suggestion
    (loadModule ./services/ihaskell.nix { condition = hostPlatform.isLinux; })
```

File: modules/misc/news.nix [hunk 12]
```diff
@@ -1076,6 +1076,13 @@ in
           support to your VSCode.
         '';
       }
+
+      {
+        time = "2019-04-27T16:41:50+00:00";
```
Comment (rycee) [inline modules/misc/news.nix]: Since the module is limited to Linux the news entry also should be:
```
condition = hostPlatform.isLinux;
```

File: modules/services/ihaskell.nix [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+      example = literalExample ''
+        ghc864
+      '';
+      description = ''
+        The Haskell compiler to use
+      '';
+    };
+
+    notebooksPath = mkOption {
+      default = "$HOME/ihaskell";
```
Comment (rycee) [inline modules/services/ihaskell.nix]: I would suggest not having a default to force the user to make a conscious decision about where they want the data. You can have this default be the example.

File: modules/services/ihaskell.nix [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+  };
+  ihaskellResult = (import (ihaskellSrc + /release.nix) {
+    compiler = cfg.compiler;
+    packages = cfg.extraPackages;
+  });
+
+in
+
+{
+  meta.maintainers = [ maintainers.rycee ];
```
[the remark this answers is not in the document]
Comment (srid) [reply modules/services/ihaskell.nix]: No problem, I can maintain it.

File: modules/services/ihaskell.nix (hunk 10 above)
Comment (srid) [reply modules/services/ihaskell.nix]: ihaskell is marked as broken in nixpkgs. I'll try to open a PR over there. This is the only change left to be done in this PR. Thanks for your comments!

Comment (vaibhavsagar) [reply modules/services/ihaskell.nix]: If it helps, I pushed a new release to Hackage the other day: http://hackage.haskell.org/package/ihaskell-0.10.0.0, and I've unblocked it in Stackage: https://github.com/commercialhaskell/stackage/pull/4503

File: modules/services/ihaskell.nix [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+        After = [ "graphical-session-pre.target" ];
+        PartOf = [ "graphical-session.target" ];
+      };
+
+      Install = {
+        WantedBy = [ "graphical-session.target" ];
+      };
+
+      Service = {
+        ExecStart = "${pkgs.runtimeShell} -c \"mkdir -p ${cfg.notebooksPath}; cd ${cfg.notebooksPath}; ${ihaskellResult}/bin/ihaskell-notebook\"";
```
Comment (rycee) [inline modules/services/ihaskell.nix]: Might be worth having
```nix
ExecStart = pkgs.writeScript "ihaskell" ''
  ${pkgs.runtimeShell}
  set -e
  mkdir -p ${cfg.notebooksPath}
  cd ${cfg.notebooksPath}
  ${ihaskellResult}/bin/ihaskell-notebook
'';
```
here. Seems a bit easier to read to me.

File: modules/services/ihaskell.nix (hunk 10 above)
Comment (rycee) [reply modules/services/ihaskell.nix]: Yep, everything looks quite nice now! I'd be happy to merge once the packaging is sorted. Thanks!

Comment (srid) [reply modules/services/ihaskell.nix]: @vaibhavsagar -  could you point me to the process for updating a haskell package (ihaskell here) in nixpkgs?

Comment (vaibhavsagar) [reply modules/services/ihaskell.nix]: It's pretty involved, I will open an issue :smile:.

Comment (vaibhavsagar) [reply modules/services/ihaskell.nix]: https://github.com/NixOS/nixpkgs/issues/60825

**Comment by srid** [comment]
I'm not using this module anymore, but others - feel free to pick it up if you are inclined.
